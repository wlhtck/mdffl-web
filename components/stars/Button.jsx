import React from 'react';
import {
  string, func, bool, oneOf,
} from 'prop-types';
import styled from 'styled-components';
import { primary, white } from '../nebulas/colors';

const Button = ({
  link, onClick, text, className, external,
}) => (
  link
    ? (
      <a
        className={className}
        href={link}
        target={external ? '_blank' : ''}
        rel={external ? 'noreferrer noopener' : ''}
      >
        {text}
      </a>
    )
    : <button className={className} onClick={onClick} type="button">{text}</button>

);

const backgroundColors = {
  primary,
  secondary: white,
};

const fontColors = {
  primary: white,
  secondary: primary,
};

Button.propTypes = {
  external: bool,
  link: string,
  onClick: func,
  text: string.isRequired,
  className: string,
  type: oneOf(['primary', 'secondary']),
};

Button.defaultProps = {
  external: false,
  link: '',
  onClick: null,
  className: '',
  type: 'primary',
};

export default styled(Button)`
  font-family: 'Raleway', sans-serif;
  font-size: 16px;
  background-color: ${({ type }) => backgroundColors[type]};
  color: ${({ type }) => fontColors[type]};
  display: inline-block;
  padding: 16px;
  border: none;
  line-height: 1;
  border-radius: 4px;
  cursor: pointer;
  text-decoration: none;
`;
