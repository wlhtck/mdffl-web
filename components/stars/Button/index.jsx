import {
  string, func, bool, oneOf,
} from 'prop-types';
import { omit } from 'lodash/fp';
import {
  compose, mapProps, componentFromProp, branch, withProps,
} from 'recompose';
import withStyles from '../../nebulas/withStyles';
import styles from './styles';

const omitProps = compose(mapProps, omit);
const withExternalLinkProps = withProps({ rel: 'noreferrer noopener', target: '_blank' });
const asButton = compose(omitProps(['href']), withProps(() => ({ component: 'button' })));
const asLink = compose(
  omitProps(['onClick']),
  withProps(() => ({ component: 'a' })),
  branch(
    ({ external }) => !!external,
    withExternalLinkProps,
  ),
);

const Button = compose(
  withStyles(styles),
  branch(
    ({ href }) => !href,
    asButton,
    asLink,
  ),
  omitProps(['external']),
)(componentFromProp('component'));

Button.propTypes = {
  children: string.isRequired,
  className: string,
  external: bool,
  onClick: func,
  type: oneOf(['primary', 'secondary']),
  url: string,
};

Button.defaultProps = {
  className: '',
  external: false,
  onClick: null,
  type: 'primary',
  url: null,
};

export default Button;
