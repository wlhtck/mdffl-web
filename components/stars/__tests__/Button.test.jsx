import React from 'react';
import 'jest-styled-components';
import { mount } from 'enzyme';
import Button from '../Button';

describe('components/starts/Button', () => {
  it('renders an interal link', () => {
    expect(mount(<Button text="Button" url="/about" />)).toMatchSnapshot();
  });

  it('renders an external link', () => {
    expect(mount(<Button text="Button" url="https://google.com" external />)).toMatchSnapshot();
  });

  it('renders a button', () => {
    expect(mount(<Button text="Button" onClick={() => {}} />)).toMatchSnapshot();
  });

  it('renders a secondary button', () => {
    expect(mount(<Button text="Button" type="secondary" />)).toMatchSnapshot();
  });
});
