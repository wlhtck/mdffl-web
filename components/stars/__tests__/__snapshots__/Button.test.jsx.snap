// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/starts/Button renders a button 1`] = `
.c0 {
  font-family: 'Raleway',sans-serif;
  font-size: 16px;
  font-weight: 800;
  background-color: #1C104E;
  color: #FFFFFF;
  display: inline-block;
  padding: 16px;
  border: none;
  line-height: 1;
  border-radius: 4px;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<Button__StyledButton
  onClick={[Function]}
  text="Button"
  type="primary"
>
  <Button
    className="c0"
    external={false}
    onClick={[Function]}
    text="Button"
    type="primary"
    url=""
  >
    <button
      className="c0"
      onClick={[Function]}
      type="button"
    >
      Button
    </button>
  </Button>
</Button__StyledButton>
`;

exports[`components/starts/Button renders a secondary button 1`] = `
.c0 {
  font-family: 'Raleway',sans-serif;
  font-size: 16px;
  font-weight: 800;
  background-color: #FFFFFF;
  color: #1C104E;
  display: inline-block;
  padding: 16px;
  border: none;
  line-height: 1;
  border-radius: 4px;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<Button__StyledButton
  text="Button"
  type="secondary"
>
  <Button
    className="c0"
    external={false}
    onClick={null}
    text="Button"
    type="secondary"
    url=""
  >
    <button
      className="c0"
      onClick={null}
      type="button"
    >
      Button
    </button>
  </Button>
</Button__StyledButton>
`;

exports[`components/starts/Button renders an external link 1`] = `
.c0 {
  font-family: 'Raleway',sans-serif;
  font-size: 16px;
  font-weight: 800;
  background-color: #1C104E;
  color: #FFFFFF;
  display: inline-block;
  padding: 16px;
  border: none;
  line-height: 1;
  border-radius: 4px;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<Button__StyledButton
  external={true}
  text="Button"
  type="primary"
  url="https://google.com"
>
  <Button
    className="c0"
    external={true}
    onClick={null}
    text="Button"
    type="primary"
    url="https://google.com"
  >
    <a
      className="c0"
      href="https://google.com"
      rel="noreferrer noopener"
      target="_blank"
    >
      Button
    </a>
  </Button>
</Button__StyledButton>
`;

exports[`components/starts/Button renders an interal link 1`] = `
.c0 {
  font-family: 'Raleway',sans-serif;
  font-size: 16px;
  font-weight: 800;
  background-color: #1C104E;
  color: #FFFFFF;
  display: inline-block;
  padding: 16px;
  border: none;
  line-height: 1;
  border-radius: 4px;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<Button__StyledButton
  text="Button"
  type="primary"
  url="/about"
>
  <Button
    className="c0"
    external={false}
    onClick={null}
    text="Button"
    type="primary"
    url="/about"
  >
    <a
      className="c0"
      href="/about"
      rel=""
      target=""
    >
      Button
    </a>
  </Button>
</Button__StyledButton>
`;
